index.mjs

```diff
@@ -146,23 +143,34 @@ ParallelHook.prototype[$map] = function(f){
   return ParallelHook(map(f)(this.hook));
 }
 
+const crash = e => Future(() => { throw e });
+
 ParallelHook.prototype[$ap] = function(mf){
   return ParallelHook(Hook(c => {
-    let f, x, b = false;
-    const mirror = cache(never);
-    const rf = mf.hook.run(_f => {
-      f = _f;
-      if (b) return map(x => mirror.resolve(x) || x)(c(f(x)));
-      b = true;
-      return mirror;
-    });
-    const rx = this.hook.run(_x => {
-      x = _x;
-      if (b) return map(x => mirror.resolve(x) || x)(c(f(x)));
-      b = true;
-      return mirror;
-    });
-    return map(head)(both(rf, rx));
+    let consume = noop;
+    const rf = mf.hook.run(f => consume !== noop ? consume (f) : (
+      Future((rej, res) => { consume = x => map(y => (res(y), y))(c(f(x))) })
+    ));
+    const rx = this.hook.run(x => consume !== noop ? consume (x) : (
+      Future((rej, res) => { consume = f => map(y => (res(y), y))(c(f(x))) })
+    ));
+    const transformation = {
+      cancel: noop,
+      context: rx.context,
+      rejected: () => never,
+      resolved: () => never,
+      toString: () => `parallelHookTransform(${rx.toString()})`,
+      run: early => {
+        const action = Object.create(transformation);
+        action.cancel = rx._interpret(
+          e => early(crash(e), action),
+          x => early(reject(x), action),
+          x => early(resolve(x), action),
+        );
+        return action;
+      },
+    };
+    return rf._transform(transformation);
```

Comment:
We solve the issue by using a different, custom, transformation (as opposed to `both`) which takes into account that the input Futures will resolve each other. We have to do this using Fluture's undocumented low-level API, as Fluture has not a single transformation that takes this use-case into account.

Reply:
I think that the original bug is for Fluture to solve - after all, we are able to create this issue using purely Fluture's public API. That will be a much more difficult issue to tackle, however, as it's an issue intrinsic to the design of Fluture's parallelism.